man/man1/logtalk_doclet.1
.TH logtalk_doclet 1 "December 29, 2024" "Logtalk 3.87.0" "Logtalk Documentation"

.SH NAME
logtalk_doclet \- Logtalk script for automating running doclets

.SH SYNOPSIS
.B logtalk_doclet -v
.br
.B logtalk_doclet -p \fIprolog\fR
[\fB-d \fIresults\fR]
[\fB-t \fItimeout\fR]
[\fB-s \fIprefix\fR]
[\fB-- \fIarguments\fR]
.br
.B logtalk_doclet -h

.SH DESCRIPTION
\fBlogtalk_doclet\fR is a script for automating running Logtalk doclets. This script runs doclets found in a directory and recursively in its sub-directories by scanning for \fIdoclet.lgt\fR and \fIdoclet.logtalk\fR source files. The script assumes that the doclets use an \fIinitialization/1\fR directive to automatically run the documentation generation goal at loading time. A non-zero exit status is returned in case of failed doclets, timed out doclets, or crashed doclets.

.SH OPTIONS
.TP
.BI \-v
Print script version.
.TP
.BI \-p
Backend Prolog compiler. Required. The following identifiers are used: B-Prolog - \fIb\fR, CxProlog - \fIcx\fR, ECLiPSe - \fIeclipse\fR, GNU Prolog - \fIgnu\fR, GNU Prolog (native code) - \fIgnunc\fR, JIProlog - \fIji\fR, XVM - \fIxvm\fR, SICStus Prolog - \fIsicstus\fR, SWI-Prolog - \fIswi\fR, SWI-Prolog with Logtalk pack - \fIswipack\fR, Tau Prolog - \fItau\fR, XSB - \fIxsb\fR, YAP - \fIyap\fR.
.TP
.BI \-d
Directory to store the doclet run logs. Default is the local \fI./logtalk_doclet_logs\fR directory.
.TP
.BI \-t
Timeout in seconds for running each doclet. Default is \fI0\fR (i.e., disabled).
.TP
.BI \-s
Path prefix to suppress when printing doclet paths. Default is the empty string.
.TP
.BI \--
Arguments to be passed to the integration script used to run the doclets (no default).
.TP
.B \-h
Provide help on script usage.

.SH "EXIT STATUS"
.TP
.B 0
All doclets succeeded. There are no timed out or crashed doclets.
.TP
.B 1
Some doclets failed. But there are no timed out or crashed doclets.
.TP
.B 3
One or more doclets timed out while executing. But there are no crashed doclets.
.TP
.B 7
One or more doclets crashed the process running them.

.SH IMPLEMENTATION NOTES
The \fIgnunc\fR backend requires creating a GNU Prolog executable that includes Logtalk named \fIgplgtnc\fR and making it available in the system \fIPATH\fR. This executable can be created using the embedding script provided in the Logtalk distribution. For example:
\fBgprolog_embed_logtalk.sh -c -d ~/collect -s none -- --new-top-level && mv ~/collect/logtalk ~/collect/gplgtnc\fR

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKUSER/library/tester_versions.lgt
Logtalk source file used for printing the Logtalk and backend Prolog compiler versions.

.SH "SEE ALSO"
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2024 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
SPDX-FileCopyrightText: 1998-2025 Paulo Moura <pmoura@logtalk.org>  
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.87.0 - January ??, 2025
=========================

Logtalk compiler and runtime
----------------------------

* ADDED: New `includes/2`, `object/3`, `protocol/3`, and `category/3`
properties to the `logtalk::loaded_file_property/2` predicate.

Documentation
-------------

* UPDATED: Notes on the Prolog integration scripts.

* FIXED: Grammar and spelling typos in scripts and tests documentation.

Tools
-----

* REMOVED: Deprecated `lgtdoc` tool `.js` scripts and `ie50.xsl` file.

* IMPROVED: Computation of the Halstead complexity metrics by considering
recursive predicate calls.

* FIXED: The `packs` tool `lgtenv.ps1` script when using the `-p` option.

Installers and installation scripts
-----------------------------------

* UPDATED: Synced the MacPorts portfile with the version used upstream.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------

* ADDED: VSCode support for finding dynamic predicate references in clauses
that assert or retract clauses for them.

* FIXED: VSCode support for finding predicate implementations when the
definition is found in an included file.

* FIXED: VSCode support for finding predicate references when the call is
found in an included file.

* FIXED: VSCode support for finding predicate declarations when the reference
is found in an included file.

* FIXED: VSCode support for navigating call hierarchies where an exception
would be thrown in some cases.

Installers and installation scripts
-----------------------------------

* REMOVED: Deprecated `logtalk_user_setup.js` script. 


3.86.0 - December 18, 2024
==========================

Logtalk compiler and runtime
----------------------------

* ADDED: Experimental `logtalk::message_prefix_file/6` hook predicate to
the message printing mechanism.

* FIXED: False left-recursion warning when compiling a grammar rule for
a multifile non-terminal that calls a local non-terminal with the same
indicator.

Prolog adapter and integration files
------------------------------------

* UPDATED: Ciao Prolog adapter file to pretend that modules are supported
to allow loading libraries and examples that require file `use_module/1`
directives.

Documentation
-------------

* ADDED: Notes on encrypting applications using XVM.

* FIXED: Multiple Handbook grammar and spelling typos.

* FIXED: Multiple man pages grammar and spelling typos.

Library
-------

* FIXED: The `os::sleep/1` predicate for XSB to also accept floats.

Tools
-----

* CHANGED: The `lgtunit` tool test compilation warning and error messages to
include the file context for better integration with other developer tools.

* CHANGED: The `diagrams` tool now uses a Unicode glyph for the magnifying
lens instead of a PNG image.

* CHANGED: The `diagrams` tool now uses a Graphviz `box` shape for Prolog
modules instead of a `tab`, which is now only used for directories and
libraries.

* REMOVED: Deprecated `lgt2svg.js` script from the `diagrams` tool.

* ADDED: Experimental `diagrams` tool support for the d2 graph language.

* UPDATED: The `diagrams` tool to support graph languages such as d2 where the
node identifiers are not global.

* UPDATED: The `diagrams` tool to also support the PearAI, Windsurf, PyCharm,
and IDEA text editors when generating diagrams with local file links.

* UPDATED: The `tutor` tool to also explain duplicated predicate/non-terminal
reference linter warnings.

* UPDATED: The `tutor` tool to call the new `logtalk::message_prefix_file/6`
hook predicate to support VSCode displaying warnings and errors augmented with
explanations.

* UPDATED: The `help` tool for the new `logtalk::message_prefix_file/6` hook
predicate.

* FIXED: The `diagrams` tool to also generate an `instance/class` caption for
metaclasses.

* FIXED: Bug in the `diagrams` tool where repeated nodes could be created for
external entities.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would miss
"includes" edges.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would show
duplicated "includes" edges.

* FIXED: The `diagrams` tool support for the graph languages to quote atoms
in predicate labels and predicate node contents when required.

* FIXED: The `diagrams` tool exclude options to allow generating external
nodes for the excluded directories, files, libraries, and entities.

* FIXED: Bug in the `lgtdoc` tool when printing a warning on missing entity
directive `info/1` key.

Examples
--------

* ADDED: New `permutations` example providing benchmarks based on list
permutations. Adapted from Prolog code written and contributed by Paul
Tarau.

* UPDATED: The `assumptions` example with another individual example.

Tests
-----

* ADDED: Additional tests for the left-recursion linter warning.

* ADDED: Additional tests for the Prolog standard `round/1` evaluable.

Installers and installation scripts
-----------------------------------

* IMPROVED: The manual installation script and MacPorts Portfile instructions
for defining the `LOGTALKHOME` and `LOGTALKUSER` environment variables.

IDEs, text editors, and syntax highlighters support

scripts/logtalk_doclet.ps1
#############################################################################
## 
##   Documentation automation script
##   Last updated on October 2, 2023
## 
##   This file is part of Logtalk <https://logtalk.org/>  
##   SPDX-FileCopyrightText: 1998-2025 Paulo Moura <pmoura@logtalk.org>
##   SPDX-License-Identifier: Apache-2.0
##   
##   Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##   
##       http://www.apache.org/licenses/LICENSE-2.0
##   
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.
## 
#############################################################################


#Requires -Version 7.3

[CmdletBinding()]
param(
	[Parameter()]
	[String]$p,
	[String]$d = "$pwd\logtalk_doclet_logs",
	[String]$s = $env:USERPROFILE,
	# disable timeouts to maintain backward compatibility
	[String]$t = 0,
	[String]$a = "",
	[Switch]$v,
	[Switch]$h
)

Function Write-Script-Version {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve
	Write-Output ($myName + " 2.5")
}

Function Run-Doclets() {
param(
	[Parameter(Position = 0)]
	[String]$path
)
	$directory = Split-Path -Path $path
	$directory_short = $directory -replace $prefix, ""
	Set-Location $d
	Write-Output '*******************************************************************************'
	Write-Output "***** Documenting $directory_short"
	$name = $directory -match '/' -replace '__'
	$doclet_exit = Run-Doclet $name $documenting_goal
	if ($doclet_exit -eq 0) {
		return 0
	} elseif ($doclet_exit -eq 124) {
		Write-Output "*****         timeout"
		Add-Content -Path $d\$name.errors -Value "LOGTALK_TIMEOUT"
	} else {
		Write-Output "*****         crash"
		Add-Content -Path $d\$name.errors -Value "LOGTALK_CRASH"
	}
	Exit 0
}

Function Run-Doclet() {
param(
	[Parameter(Position = 0)]
	[String]$name,
	[Parameter(Position = 1)]
	[String]$goal
)
	$results_file = Join-Path $d ($name + ".results")
	$errors_file  = Join-Path $d ($name + ".errors")
	if ($t -ne 0) {
		& $timeout_command $t $logtalk $logtalk_option $goal $a > $results_file 2> $errors_file
	} else {
		& $logtalk $logtalk_option $goal $a > $results_file 2> $errors_file
	}
	return $?
}

Function Write-Usage-Help() {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve 

	Write-Output ""
	Write-Output "This script automates running doclets found on the current directory and recursively"
	Write-Output "in its sub-directories by scanning for doclet.lgt and doclet.logtalk source files. In"
	Write-Output  "case of failed doclets or doclet errors, this script returns a non-zero exit code."
	Write-Output ""
	Write-Output "Usage:"
	Write-Output ("  " + $myName + " -p prolog [-d results] [-t timeout] [-s prefix] [-a arguments]")
	Write-Output ("  " + $myName + " -v")
	Write-Output ("  " + $myName + " -h")
	Write-Output ""
	Write-Output "Required arguments:"
	Write-Output "  -p backend Prolog compiler"
	Write-Output "     (possible values are b, ciao, cx, eclipse, gnu, gnunc, ji, xvm, sicstus, swi, swipack, tau, trealla, xsb, and yap)"
	Write-Output ""
	Write-Output "Optional arguments:"
	Write-Output "  -d directory to store the doclet logs (default is ./logtalk_doclet_logs)"
	Write-Output ("  -t timeout in seconds for running each doclet (default is " + $t + "; i.e. disabled)")
	Write-Output ("  -s suppress path prefix (default is " + $s + ")")
	Write-Output "  -a arguments wrapped as a string to be passed to the integration script used to run the doclets (no default)"
	Write-Output "  -v print version"
	Write-Output "  -h help"
	Write-Output ""
}

Function Check-Parameters() {

	if ($v -eq $true) {
		Write-Script-Version
		Exit
	}

	if ($h -eq $true) {
		Write-Usage-Help
		Exit
	}

	if ($p -eq "") {
		Write-Output ("Error! Backend Prolog compiler not specified!")
		Write-Usage-Help
		Exit 1
	} elseif ($p -eq "b") {
		$script:prolog='B-Prolog'
		$script:logtalk="bplgt"
		$script:logtalk_option="-g"
	} elseif ($p -eq "ciao") {
		$script:prolog='Ciao Prolog'
		$script:logtalk="ciaolgt"
		$script:logtalk_option="-e"
	} elseif ($p -eq "cx") {
		$script:prolog='CxProlog'
		$script:logtalk="cxlgt"
		$script:logtalk_option="--goal"
	} elseif ($p -eq "eclipse") {
		$script:prolog='ECLiPSe'
		$script:logtalk="eclipselgt"
		$script:logtalk_option="-e"
	} elseif ($p -eq "gnu") {
		$script:prolog='GNU Prolog'
		$script:logtalk="gplgt"
		$script:logtalk_option="--query-goal"
	} elseif ($p -eq "gnunc") {
		$script:prolog = 'GNU Prolog (native code)'
		$script:logtalk = "gplgtnc"
		$script:logtalk_option="--query-goal"
	} elseif ($p -eq "ji") {
		$script:prolog='JIProlog'
		$script:logtalk="jiplgt"
		$script:logtalk_option="-n -g"
	} elseif ($p -eq "xvm") {
		$script:prolog='XVM'
		$script:logtalk="xvmlgt"
		$script:logtalk_option="-g"
		$script:dot="."
	} elseif ($p -eq "sicstus") {
		$script:prolog='SICStus Prolog'
		$script:logtalk="sicstuslgt"
		$script:logtalk_option="--goal"
		$script:dot="."
	} elseif ($p -eq "swi") {
		$script:prolog='SWI-Prolog'
		$script:logtalk="swilgt"
		$script:logtalk_option="-g"
	} elseif ($p -eq "swipack") {
		$script:prolog='SWI-Prolog'
		$script:logtalk="swipl"
		$script:logtalk_option="-g"
	} elseif ($p -eq "tau") {
		$script:prolog='Tau Prolog'
		$script:logtalk="taulgt"
		$script:logtalk_option="-g"
	} elseif ($p -eq "trealla") {
		$script:prolog='Trealla Prolog'
		$script:logtalk="tplgt"
		$script:logtalk_option="-g"
	} elseif ($p -eq "xsb") {
		$script:prolog='XSB'
		$script:logtalk="xsblgt"
		$script:logtalk_option="-e"
		$script:dot="."
	} elseif ($p -eq "yap") {
		$script:prolog='YAP'
		$script:logtalk="yaplgt"
		$script:logtalk_option="-g"
	} else {
		Write-Output ("Error! Unsupported backend Prolog compiler: " + "$p")
		Write-Usage-Help
		Exit
	}

	if ($p -eq "swipack") {
		$script:versions_goal = ("use_module(library(logtalk))," + $versions_goal)
	}

}

###################### here it starts ############################ 

# documenting goals

$versions_goal = ("logtalk_load(library(tester_versions)),halt" + $dot)
$documenting_goal = ("logtalk_load([doclet(loader),doclet]),halt" + $dot)

# default argument values

$base = $pwd
$dot = ""
$prolog = 'SWI-Prolog'
$logtalk = "swilgt"
$logtalk_option = "-g"

Check-Parameters

New-Item -Path $d -ItemType directory -Force > $null

if (Test-Path $d\*.results) {
	Remove-Item -Path $d\*.results -Force
}
if (Test-Path $d\*.errors) {
	Remove-Item -Path $d\*.errors -Force
}
if (Test-Path $d\errors.all) {
	Remove-Item -Path $d\errors.all -Force
}
if (Test-Path $d\tester_versions.txt) {
	Remove-Item -Path $d\tester_versions.txt -Force
}

$start_date = Get-Date -Format "yyyy-MM-dd-HH:mm:ss"

Write-Output '*******************************************************************************'
Write-Output ("***** Batch documentation processing started @ " + $start_date)
$tester_versions_file = Join-Path $d tester_versions.txt
& $logtalk $logtalk_option (" `"" + $versions_goal + "`"") > $tester_versions_file 2> $null
Select-String -Path $d\tester_versions.txt -Pattern "Logtalk version:" -SimpleMatch -Raw
(Select-String -Path $d\tester_versions.txt -Pattern "Prolog version:"  -SimpleMatch -Raw) -replace "Prolog", $prolog

$doclets=0
Get-ChildItem -Path . -Include doclet.lgt doclet.logtalk -Recurse | 
Foreach-Object {
	$doclets++
	Run-Doclets $_.FullName
}

Push-Location $d

$timeouts = (Get-ChildItem -Path . -Filter *.errors  | Select-String -Pattern 'LOGTALK_TIMEOUT').count
$crashes =  (Get-ChildItem -Path . -Filter *.errors  | Select-String -Pattern 'LOGTALK_CRASH').count
$failures = (Get-ChildItem -Path . -Filter *.results | Select-String -Pattern 'failed').count

Write-Output "*******************************************************************************"
Write-Output "***** Compilation errors/warnings and failed doclets"
Write-Output "*******************************************************************************"
(Get-ChildItem -Path . -Filter *.results  | Select-String -Pattern 'syntax_error' -SimpleMatch -Raw) -replace '.results', '' | Tee-Object -FilePath errors.all -Append
(Get-ChildItem -Path . -Filter *.errors   | Select-String -Pattern 'syntax_error' -SimpleMatch -Raw) -replace '.errors', ''  | Tee-Object -FilePath errors.all -Append
(Get-ChildItem -Path . -Filter *.results  | Select-String -Pattern '!     ' -SimpleMatch -Raw) -replace '.results', '' | Tee-Object -FilePath errors.all -Append
(Get-ChildItem -Path . -Filter *.errors   | Select-String -Pattern '!     ' -SimpleMatch -Raw) -replace '.errors', ''  | Tee-Object -FilePath errors.all -Append
(Get-ChildItem -Path . -Filter *.results  | Select-String -Pattern '*     ' -SimpleMatch -Raw) -replace '.results', '' | Tee-Object -FilePath errors.all -Append
(Get-ChildItem -Path . -Filter *.errors   | Select-String -Pattern '*     ' -SimpleMatch -Raw) -replace '.errors', ''  | Tee-Object -FilePath errors.all -Append
Write-Output "*******************************************************************************"
Write-Output "***** Timeouts"
Write-Output "*******************************************************************************"
((Get-ChildItem -Path . -Filter *.errors | Select-String -Pattern 'LOGTALK_TIMEOUT' -SimpleMatch -Raw) -replace '\\', '__') -replace $prefix, ''
Write-Output "*******************************************************************************"
Write-Output "***** Crashes"
Write-Output "*******************************************************************************"
((Get-ChildItem -Path . -Filter *.errors | Select-String -Pattern 'LOGTALK_CRASH' -SimpleMatch -Raw) -replace '\\', '__') -replace $prefix, ''
Write-Output "*******************************************************************************"
Write-Output "***** Failures"
Write-Output "*******************************************************************************"
Get-ChildItem -Path . -Filter *.results | Select-String -Pattern 'failed' -SimpleMatch -Raw
Write-Output "*******************************************************************************"
Write-Output ("***** " + $doclets + " doclets: " + $timeouts + " timeouts, " + $crashes + " crashes, " + $failures + " failures")
Write-Output "*******************************************************************************"

$end_date = Get-Date -Format "yyyy-MM-dd-HH:mm:ss"

Write-Output ("***** Batch documentation processing ended @ " + $end_date)
Write-Output '*******************************************************************************'

Pop-Location

if ($crashes -gt 0) {
	Exit 7
} elseif ($timeouts -gt 0) {
	Exit 3
} elseif ($failures -gt 0) {
	Exit 1
} else {
	Exit 0
}
